File: testdata/rules/MCPG001/safe/files.ts
import { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import { z } from "zod";
import * as fs from "node:fs/promises";
import * as path from "node:path";

const ROOT = path.resolve("./workspace");
const server = new McpServer({ name: "fs", version: "1.0.0" });

server.tool(
  "save_file",
  "Save a file inside the workspace",
  { filePath: z.string(), data: z.string() },
  async ({ filePath, data }) => {
    const target = path.resolve(ROOT, filePath);
    const rel = path.relative(ROOT, target);
    if (rel.startsWith("..") || path.isAbsolute(rel)) {
      throw new Error("path escapes the workspace");
    }
    await fs.writeFile(target, data);
    return { content: [{ type: "text", text: "ok" }] };
  }
);

File: testdata/rules/MCPG001/safe/files.py
from pathlib import Path

from mcp.server.fastmcp import FastMCP

ROOT = Path("/srv/notes").resolve()
mcp = FastMCP("notes")


@mcp.tool()
def write_note(path: str, content: str) -> str:
    """Write a note inside the notes directory."""
    target = (ROOT / path).resolve()
    if not target.is_relative_to(ROOT):
        raise ValueError("path escapes the notes directory")
    target.write_text(content, encoding="utf-8")
    return "saved"


@mcp.tool()
def list_notes() -> list[str]:
    """List notes."""
    return [p.name for p in ROOT.iterdir()]
